### ISSUE TITLE
Issues with periodic boundaries

### ISSUE THREAD

**Issue by natschil** [JuliaMath/Interpolations.jl#326]
It seems like there is an issue somewhere with the periodic boundary conditions. If I approximate e.g. the sin-function in 1D, the result has errors near the boundary. 

The following is a minimal (non) working example:
```
using Interpolations
xs = range(0,stop=2π,length=100)
sinx = sin.(xs)
itp = interpolate(sinx,BSpline(Cubic(Periodic(OnGrid()))))
itp2 = scale(itp,xs)
itp3 = extrapolate(itp2,Periodic())
plot(range(2π-0.1,stop=2π+.1,length=500),x->itp3(x),label="Interpolated")
plot!(range(2π-0.1,stop=2π+.1,length=500),x->sin(x),label="Exact")
```

**Comment by natschil** [comment]
[output.pdf](https://github.com/JuliaMath/Interpolations.jl/files/3321315/output.pdf)

**Comment by natschil** [comment]
If I do the same thing but without the point at 2\pi (i.e. xs = range(0,stop=2π,length=100)[1:(end-1)]), the result looks worse:
![output](https://user-images.githubusercontent.com/3118879/60028122-105ce280-969f-11e9-81db-9372eb68ad5d.png)

**Comment by natschil** [comment]
We can take the example from https://github.com/JuliaMath/Interpolations.jl/issues/201 which is very closely related, and plot the sin function next to it:

```f(x) = sin(x)
xs = range(0, stop=2pi, length=10)[1:end-1] # note that we don't want the overlapping point at 2pi
A = f.(xs);
itp = interpolate(A, BSpline(Linear()), OnGrid()) # interpolate linearly between the data points
stp = scale(itp, xs) # re-scale to the actual domain
etp = extrapolate(stp, Periodic()) # extrapolate periodically

xs_long = range(-2pi, stop=4pi, length=100)
A_long = [etp[x] for x in xs_long]
p = plot(xs, A, seriestype=:scatter, label="Data")
plot!(p, xs_long, A_long, label="Extrapolated")
plot!(p,xs_long,sin,label="sin(x)")
```
![output](https://user-images.githubusercontent.com/3118879/60058670-99990700-96e9-11e9-94c0-09faf0e71333.png)

It seems like in this case we *do* need the additional
![output2](https://user-images.githubusercontent.com/3118879/60058698-bd5c4d00-96e9-11e9-9f5c-d32e22dd796c.png)
 point at 2\pi:

Note that even if this point is included there is an issue with Cubic() interpolations, and ideally one would want to not need that point as one "knows" it already.

**Comment by briochemc** [comment]
Just for the record, here is another workaround using ImageFiltering.jl's `padarray`, that requires a few intermediate steps. The idea below was to enforce the periodicity on the argument of `LinearInterpolation`, but since I'm not entirely sure how to do that cleanly, I just allocated a whole new "periodically extended" array as the argument:

```julia
xs = range(0, 2π, length=100)
xmids = (xs[2:end] + xs[1:end-1]) / 2 # The circular padding has no option for repeated end-points AFAIK
sinx = sin.(xmids)
sinx_pad = padarray(sinx, Pad(:circular, 200)) # enforce periodicity of the function
itp_x = LinearInterpolation(1:length(xmids), xmids, extrapolation_bc = Line()) # extrapolate x values linearly
# Now use standard interpolation on extended, periodic input
itp = LinearInterpolation(itp_x[-50:150], sinx_pad[-50:150])
plot(range(2π-0.1, 2π+0.1, length=500), x->itp(x), lab="Interpolated")
plot!(range(2π-0.1, 2π+0.1, length=500), sin, lab="Exact")
```

Maybe something like that could be implemented under the hood of Interpolations.jl?

**Comment by mkitti** [comment]
Also see https://github.com/roflmaostc/FFTResampling.jl

**Comment by jaemolihm** [comment]
Another workaround for periodic linear interpolation without additional points is as follows. The motivation is that one only needs to modify `lbound` and `ubound` to get what we want.

Below, `ConstantPeriodic` and `LinearPeriodic` are similar to `Constant` and `Linear`, but are a subtype of `DegreeBC`. Functions (`positions`, `value_weights`, ...) are defined in the same way as `Constant` and `Linear`. The only exception is `positions(::LinearPeriodic, ax::AbstractUnitRange{<:Integer}, x)`, where one needs to return `(modrange(fi, ax), modrange(fi+1, ax))` instead of `fi` to use `WeightedArbIndex` instead of `WeightedAdjIndex`.

I have little experience with `Interpolations.jl`, so I am not sure this workaround applies to `LinearInterpolation` itself (instead of Linear BSpline) or not.

```julia
using Interpolations
using PyPlot
using Interpolations: BoundaryCondition, DegreeBC, GridType
using Interpolations: roundbounds, fast_trunc, modrange

# Define ConstantPeriodic
struct ConstantPeriodic{BC<:BoundaryCondition} <: DegreeBC{0}
    bc::BC
end
(deg::ConstantPeriodic)(gt::GridType) = ConstantPeriodic(deg.bc(gt))

function Interpolations.positions(c::ConstantPeriodic, ax, x)  # discontinuity occurs at half-integer locations
    xm = roundbounds(x, ax)
    δx = x - xm
    fast_trunc(Int, xm), δx
end
Interpolations.value_weights(::ConstantPeriodic, δx) = (1,)
Interpolations.gradient_weights(::ConstantPeriodic, δx) = (0,)
Interpolations.hessian_weights(::ConstantPeriodic, δx) = (0,)
Interpolations.padded_axis(ax::AbstractUnitRange, ::BSpline{<:ConstantPeriodic}) = ax

# Define LinearPeriodic
struct LinearPeriodic{BC<:BoundaryCondition} <: DegreeBC{1}
    bc::BC
end
(deg::LinearPeriodic)(gt::GridType) = LinearPeriodic(deg.bc(gt))

function Interpolations.positions(::LinearPeriodic, ax::AbstractUnitRange{<:Integer}, x)
    f = floor(x)
    # When x == last(ax) we want to use the x-1, x pair
    f = ifelse(x == last(ax), f - oneunit(f), f)
    fi = fast_trunc(Int, f)
    return (modrange(fi, ax), modrange(fi+1, ax)), x-f # Only line different from linear.jl
end
Interpolations.value_weights(::LinearPeriodic, δx) = (1-δx, δx)
Interpolations.gradient_weights(::LinearPeriodic, δx) = (-oneunit(δx), oneunit(δx))
Interpolations.hessian_weights(::LinearPeriodic, δx) = (zero(δx), zero(δx))
Interpolations.padded_axis(ax::AbstractUnitRange, ::BSpline{<:LinearPeriodic}) = ax

begin
    f(x) = sin(2π * x)
    df(x) = 2π * cos(2π * x)
    N = 5
    x = range(0, 1, length=N+1)[1:end-1]
    y = f.(x)
    degs = reshape([Constant(), ConstantPeriodic(Periodic(OnCell())),
                    Linear(), LinearPeriodic(Periodic(OnCell())),
                    Cubic(Periodic(OnGrid())), Cubic(Periodic(OnCell()))],
                    2, 3)
    fig, axes = subplots(size(degs)..., figsize=(12, 8))
    for (i, deg) in enumerate(degs)
        sca(axes[i])
        itp = interpolate(y, BSpline(deg));
        zero_to_one = range(0, 1, length=N+1)[1:end-1]
        stp = scale(itp, zero_to_one);
        etp = extrapolate(stp, Periodic());
        xx = range(-0.5, 2.5, step=0.001);
        plot(xx, etp.(xx), "C0-", label="Interpolated")
        plot(x, y, "ko", label="Data")
        plot(xx, f.(xx), "k--", label="Exact")
        # # Plot derivatives
        # if i >= 3
        #     plot(xx, Interpolations.gradient.(Ref(etp), xx) ./ 5, "C1-")
        #     plot(xx, df.(xx) ./ 5, "y--")
        # end
        title("$deg")
        if i == 1
            legend(loc="lower right")
        end
    end
    display(gcf()); clf()
    close(fig)
end
```

The first two columns in the lower row are the results of this workaround.

![image](https://user-images.githubusercontent.com/32537613/119635058-476d3d00-be4e-11eb-9b64-7080711d787b.png)

**Comment by mkitti** [comment]
I did find a bug in periodic boundaries in #419 and implemented a fix in 
https://github.com/JuliaMath/Interpolations.jl/commit/f74958055a5f3618ef43f13271982baf518ee647 . I wonder if that also fixes this issue. Could someone try master.

**Comment by mkitti** [comment]
> The idea is that one only needs to call different `lbound` and `ubound` implementations to get what we want, and this can be achieved by subtyping `DegreeBC` instead of `Degree`.

This probably could be a pull request although I would learn towards creating a `Linear{Periodic}` etc.

**Comment by jaemolihm** [comment]
> I did find a bug in periodic boundaries in #419 and implemented a fix in f749580 . I wonder if that also fixes this issue. Could someone try master.

I used the current master (`dd17e230c9cffea15d49912163094bd0077bdee4`) to generate above figures. It fixes the problem of cubic spline of #419. But not the problem of constant and linear spline.

> This probably could be a pull request although I would learn towards creating a Linear{Periodic} etc.

The problem I had was that `Linear <: Degree`, but I need `LinearPeriodic <: DegreeBC <: Degree`. I agree that `LinearPeriodic` is not a clean solution, so I called it a workaround. I will try making `Linear` and `Constant` a subtype of `DegreeBC` (I was worried if it will have any side effects), and if it works, create a PR. If you have any other suggestions, I will be happy to try.
